### PR TITLE
3960 when creating a resource patient where patientmetasource is provided with a hash the hash value is overwritten

### PR DESCRIPTION
**Pull request by MykolaMedynskyiSCDR**
What was done:
-Whether the submitted request id will be preserved now depends on the configuration.
-Added new test to check request id replacement.
-Added change log.

### CODE REVIEW

File: hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/api/config/DaoConfig.java [hunk 1]
```diff
@@ -2911,6 +2916,27 @@ public void setUpdateWithHistoryRewriteEnabled(boolean theUpdateWithHistoryRewri
...
+	 * @since 6.2.0
+	 */
+	public boolean isPreserveRequestIdInResourceBody() {
+		return myPreserveRequestIdInResourceBody;
+	}
+
+	/**
+	 * This setting indicate whether a providedResource.meta.source requestID (source#requestID)
+	 * should be preserved or overwritten.
+	 * Default is false.
```
Comment (epeartree) [inline hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/api/config/DaoConfig.java:2932]: This is good.  Still, i think we could provide further clarifications like "Default is false ie. a client provided requestId will be overwritten."

File: hapi-fhir-jpaserver-test-utilities/src/test/java/ca/uhn/fhir/jpa/provider/r4/ResourceProviderR4Test.java [hunk 2]
```diff
@@ -214,6 +215,9 @@ public class ResourceProviderR4Test extends BaseResourceProviderR4Test {
 	@Autowired
 	private ISearchDao mySearchEntityDao;
 
+	@Autowired
```
Comment (epeartree) [inline hapi-fhir-jpaserver-test-utilities/src/test/java/ca/uhn/fhir/jpa/provider/r4/ResourceProviderR4Test.java:218]: Please remove the historyProvenanceDao as it is no longer needed.

File: hapi-fhir-jpaserver-test-utilities/src/test/java/ca/uhn/fhir/jpa/provider/r4/ResourceProviderR4Test.java [hunk 3]
```diff
@@ -7082,6 +7090,231 @@ public void testUpdateHistoryRewriteWithIdNoIdPart() {
 		}
 	}
 
+	@Test
+	public void createResource_withPreserveRequestIdEnabled_RequestIdIsPreserved(){
```
Comment (epeartree) [inline hapi-fhir-jpaserver-test-utilities/src/test/java/ca/uhn/fhir/jpa/provider/r4/ResourceProviderR4Test.java:7094]: Please lower case my mistake (RequestIdIsPreserved to requestIdIsPrserved)

File: hapi-fhir-jpaserver-test-utilities/src/test/java/ca/uhn/fhir/jpa/provider/r4/ResourceProviderR4Test.java [hunk 4]
```diff
@@ -7082,6 +7090,231 @@ public void testUpdateHistoryRewriteWithIdNoIdPart() {
...
+
+		Patient returnedPatient = myClient.read().resource(Patient.class).withId(iIdType).execute();
+
+		String returnedPatientMetaSource = returnedPatient.getMeta().getSource();
+
+		assertEquals(expectedMetaSource, returnedPatientMetaSource);
+	}
+
+	@Test
+	public void createResource_withPreserveRequestIdDisabled_RequestIdIsOverwritten(){
```
Comment (epeartree) [inline hapi-fhir-jpaserver-test-utilities/src/test/java/ca/uhn/fhir/jpa/provider/r4/ResourceProviderR4Test.java:7117]: this test and the one above can be combined with the use of a @MethodSource but it's no big deal at all.

File: hapi-fhir-jpaserver-test-utilities/src/test/java/ca/uhn/fhir/jpa/provider/r4/ResourceProviderR4Test.java [hunk 5]
```diff
@@ -7082,6 +7090,231 @@ public void testUpdateHistoryRewriteWithIdNoIdPart() {
...
+
+		Patient returnedPatient = (Patient) results.getEntry().get(0).getResource();
+		String returnedPatientMetaSource = returnedPatient.getMeta().getSource();
+
+		assertEquals(1, results.getEntry().size());
+		assertEquals(expectedSourceUrl, returnedPatientMetaSource);
+	}
+
+	@Test
+	public void searchResource_bySourceAndCustomRequestIdWithPreserveRequestIdDisabled_fails(){
```
Comment (epeartree) [inline hapi-fhir-jpaserver-test-utilities/src/test/java/ca/uhn/fhir/jpa/provider/r4/ResourceProviderR4Test.java:7169]: please remove 'Custom'

File: hapi-fhir-jpaserver-test-utilities/src/test/java/ca/uhn/fhir/jpa/provider/r4/ResourceProviderR4Test.java [hunk 6]
```diff
@@ -7082,6 +7090,231 @@ public void testUpdateHistoryRewriteWithIdNoIdPart() {
...
+			.search()
+			.byUrl(ourServerBase + "/Patient?_source=%231234567890123456:)")
+			.returnBundle(Bundle.class)
+			.execute();
+
+		assertEquals(0, results.getEntry().size());
+	}
+
+	@Test
+	public void searchResource_bySourceAndWrongCustomRequestIdWithPreserveRequestIdEnabled_fails() {
```
Comment (epeartree) [inline hapi-fhir-jpaserver-test-utilities/src/test/java/ca/uhn/fhir/jpa/provider/r4/ResourceProviderR4Test.java:7299]: remove 'custom'

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/BaseHapiFhirDao.java [hunk 7]
```diff
@@ -1200,13 +1199,14 @@ public <R extends IBaseResource> R toResource(Class<R> theResourceType, IBaseRes
...
 		// 6. Handle source (provenance)
+//		if (!myConfig.isPreserveRequestIdInResourceBody() && isNotBlank(provenanceSourceUri) && provenanceSourceUri.contains("#") && !provenanceSourceUri.endsWith("#")) {
+//				MetaUtil.setSource(myContext, retVal, provenanceSourceUri);
+//		} else if (isNotBlank(provenanceRequestId) || isNotBlank(provenanceSourceUri)) {
 		if (isNotBlank(provenanceRequestId) || isNotBlank(provenanceSourceUri)) {
 			String sourceString = cleanProvenanceSourceUri(provenanceSourceUri)
 				+ (isNotBlank(provenanceRequestId) ? "#" : "")
 				+ defaultString(provenanceRequestId);
-
 			MetaUtil.setSource(myContext, retVal, sourceString);
```
Comment (epeartree) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/BaseHapiFhirDao.java:1209]: please add breathing lines before and after MetaUtil.setSource

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/BaseHapiFhirDao.java [hunk 8]
```diff
@@ -1643,6 +1643,9 @@ private void createHistoryEntry(RequestDetails theRequest, IBaseResource theReso
 			}
 		}
 
+		String requestId = getRequestId(theRequest, source);
```
Comment (epeartree) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/BaseHapiFhirDao.java:1646]: this is good to encapsulate the logic in a properly named method.  really adds clarity to the code.
please move the line to where requestId was previously defined.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/BaseHapiFhirDao.java [hunk 9]
```diff
@@ -1660,6 +1663,16 @@ private void createHistoryEntry(RequestDetails theRequest, IBaseResource theReso
 		}
 	}
 
+	private String getRequestId(RequestDetails theRequest, String source) {
+		if (myConfig.isPreserveRequestIdInResourceBody()) {
```
Comment (epeartree) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/BaseHapiFhirDao.java:1667]: This is one hard logic to read.  i'm turning green.  cant' figure out what's going on unless i write my own unit test, which i did.

let's try replacing the entire logic withing the if statement with
return StringUtils.substringAfter(theSource, "#");


Re-inventing the wheel is innefficient and introduces errors.  when ever i have to do some type of manipulation that seems trivial or common, i'll try to find a utility method to do it for me.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/BaseHapiFhirDao.java [hunk 10]
```diff
@@ -1660,6 +1663,16 @@ private void createHistoryEntry(RequestDetails theRequest, IBaseResource theReso
 		}
 	}
 
+	private String getRequestId(RequestDetails theRequest, String source) {
```
Comment (epeartree) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/BaseHapiFhirDao.java:1666]: change String source to String theSource.  member properties always start with 'my' while method parameters start with 'the'

File: hapi-fhir-jpaserver-test-utilities/src/test/java/ca/uhn/fhir/jpa/provider/r4/ResourceProviderR4Test.java [hunk 11]
```diff
@@ -235,6 +239,10 @@ public void after() throws Exception {
 
 		myClient.unregisterInterceptor(myCapturingInterceptor);
 		myDaoConfig.setUpdateWithHistoryRewriteEnabled(false);
+		myDaoConfig.setPreserveRequestIdInResourceBody(false);
+
+		myDaoConfig.setStoreMetaSourceInformation(DaoConfig.StoreMetaSourceInformationEnum.SOURCE_URI_AND_REQUEST_ID);
```
Comment (epeartree) [inline hapi-fhir-jpaserver-test-utilities/src/test/java/ca/uhn/fhir/jpa/provider/r4/ResourceProviderR4Test.java:244]: i've added this setStoreMeta... to ensure that the associated property was indeed source_uri_and_request_id.  i don't think it is needed.  i'd try to remove it, re-run all tests in the class and see if we broke something.  and most likely we wont.

File: hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/6_2_0/3960-when-creating-a-resource-patient-where-patientmetasource-is-provided-with-a-hash-the-hash-value-is-overwritten.yaml [hunk 12]
```diff
@@ -0,0 +1,5 @@
+---
+type: fix
+issue: 3960
+jira: SMILE-4679
+title: "Previously, when the source in meta was specified with id, the id was replaced. This has been corrected."
```
Comment (epeartree) [inline hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/6_2_0/3960-when-creating-a-resource-patient-where-patientmetasource-is-provided-with-a-hash-the-hash-value-is-overwritten.yaml:5]: now that we are experts with the meta resource, let's make this clearer by adding more meat to the bone:
Previously, when a client would provide a requestId within the source uri of a Meta.source, the provided requestId would get discarded and replaced by an id generated by the system.  This has been corrected.

**Review by epeartree — CHANGES_REQUESTED** [review]
Please have a look at the comments.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/BaseHapiFhirDao.java (hunk 8 above)
Comment (MykolaMedynskyiSCDR) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/BaseHapiFhirDao.java:1644]: Now requestId depends on source, and in place where it was previously defined the source does not yet have its value.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/BaseHapiFhirDao.java [hunk 13]
```diff
@@ -1663,12 +1661,10 @@ private void createHistoryEntry(RequestDetails theRequest, IBaseResource theReso
...
 
-	private String getRequestId(RequestDetails theRequest, String source) {
+	private String getRequestId(RequestDetails theRequest, String theSource) {
 		if (myConfig.isPreserveRequestIdInResourceBody()) {
-			String[] splitSource = source.split("#", 2);
-			if (splitSource.length == 2) {
-				return splitSource[1].equals("") ? null : splitSource[1] ;
-			}
+			String requestId = StringUtils.substringAfter(theSource, "#");
+			return requestId.equals("") ? null : requestId;
```
Comment (epeartree) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/BaseHapiFhirDao.java:1667]: can we really have EMPTY sting as a value returned by substringAfter

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/BaseHapiFhirDao.java [hunk 14]
```diff
@@ -205,7 +204,7 @@
 	@Autowired
 	protected IForcedIdDao myForcedIdDao;
 	@Autowired
-	protected IResourceProvenanceDao myResourceProvenanceDao;
+	protected IResourceHistoryTableDao myResourceProvenanceDao;
```
[the remark this answers is not in the document]
Comment (epeartree) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/BaseHapiFhirDao.java:207]: The name was change for the interface to better reflect it's nature, that is, to be a JpaRepository of ResourceHistoryProvenanceEntity.  The name could be further refined to 'IResourceHistoryProvenanceDao'.

Modification will be made.

[the remark this answers is not in the document]
Comment (StevenXLi) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/BaseHapiFhirDao.java:207]: I understand reasoning behind the change to the interface's name. But is the change here specifically intentional? Are we using the History table instead?

[the remark this answers is not in the document]
Comment (MykolaMedynskyiSCDR) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/BaseHapiFhirDao.java]: We do not use it so I removed it.

**Review by epeartree — APPROVED** [review]

**Review by StevenXLi — APPROVED** [review]

**Pull request merged**